### PR TITLE
[DAGCombiner] Ensure poison-generating flags are stripped in `freeze` op

### PR DESCRIPTION
**Pull request by antoniofrighetto**
When combining `freeze` in DAG, poison-generating flags were previously not fully removed as intended. This change explicitly drops these flags from `freeze` operand, ensuring a fixed, non-poisonous value at this point in computation.

Previous buildbot failures: https://lab.llvm.org/buildbot/#/builders/145/builds/2872/steps/11/logs/stdio, https://lab.llvm.org/buildbot/#/builders/95/builds/5659/steps/11/logs/stdio.

### CODE REVIEW

File: llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll [hunk 1]
```diff
@@ -0,0 +1,40 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=arm-linux-gnueabi -mcpu=arm1022e -mattr=armv5te -o - %s | FileCheck %s
+
+; Ensure poison-generating flags are stripped by the time a freeze operand is visited.
+
+@g_ptr = global ptr null, align 4
+
+define ptr @drop_flags(i32 noundef %numentries, i64 %cond, i64 %arg) local_unnamed_addr #0 {
```
Comment (arsenm) [inline llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll:8]: Don't need local_unnamed_addr, and `#0` is dead

File: llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll [hunk 2]
```diff
@@ -0,0 +1,40 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=arm-linux-gnueabi -mcpu=arm1022e -mattr=armv5te -o - %s | FileCheck %s
```
Comment (arsenm) [inline llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll:2]: Pretty sure this would need to be +armv5te, does this print a parse warning?

Comment (antoniofrighetto) [reply llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll]: It's quite weird as I didn't get any warning, this definitely needs to be fixed. Dropped it as turned to be not needed (gets exercized anyways).

**Comment by RKSimon** [comment]
> When combining freeze in DAG, poison-generating flags were previously not fully removed as intended. This change explicitly drops these flags from freeze operand, ensuring a fixed, non-poisonous value at this point in computation.

Sorry if I've missed something (long day), but where was it intended that the poison flags should be removed?

**Comment by antoniofrighetto** [comment]
> > When combining freeze in DAG, poison-generating flags were previously not fully removed as intended. This change explicitly drops these flags from freeze operand, ensuring a fixed, non-poisonous value at this point in computation.
> 
> Sorry if I've missed something (long day), but where was it intended that the poison flags should be removed?

I might be missing something, but if I read this correctly, the line following the comment would be supposed to strip the flags from the current node (which doesn't carry any flags), and getNode doesn't actually drop any flags. To be correct, it seems that no flags are set in FlagsInserter in the current DAG, except that `samesign` is there.

File: llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+entry:
+  %cmp4 = icmp samesign ult i64 %cond, %arg
+  %conv6 = trunc nuw i64 %cond to i32
+  %spec.select = select i1 %cmp4, i32 %conv6, i32 %numentries
+  %spec.select.fr = freeze i32 %spec.select
+  %cmpz = icmp eq i32 %spec.select.fr, 0
+  br i1 %cmpz, label %bb.end, label %bb.false
+
+bb.false:                                       ; preds = %entry
+  %2 = tail call range(i32 0, 33) i32 @llvm.ctlz.i32(i32 %spec.select.fr, i1 true)
```
[the remark this answers is not in the document]
Comment (antoniofrighetto) [reply llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll]: Test simplified.

File: llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+entry:
+  %cmp4 = icmp samesign ult i64 %cond, %arg
+  %conv6 = trunc nuw i64 %cond to i32
+  %spec.select = select i1 %cmp4, i32 %conv6, i32 %numentries
+  %spec.select.fr = freeze i32 %spec.select
+  %cmpz = icmp eq i32 %spec.select.fr, 0
+  ret i1 %cmpz
+}
+
+declare i32 @llvm.ctlz.i32(i32, i1)
```
Comment (dtcxzyw) [inline llvm/test/CodeGen/ARM/dagcombine-drop-flags-freeze.ll:25]: Unused.

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp [hunk 5]
```diff
@@ -16151,6 +16151,7 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
                              SVN->getMask());
   } else {
     // NOTE: this strips poison generating flags.
+    N0->dropFlags(SDNodeFlags::PoisonGeneratingFlags);
```
Comment (dtcxzyw) [inline llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16154]: I don't think it is a correct fix. The following `getNode` creates a new copy without flags. However, it is CSEed to N0 in `SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, SDVTList VTList, ArrayRef<SDValue> Ops, const SDNodeFlags Flags)`.

We should intersect flags after CSE here:
https://github.com/llvm/llvm-project/blob/92daad2eac587cb0592de019cd5f6cbb7c42bb78/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L10527-L10528

I will post a fix later.

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp (hunk 5 above)
Comment (antoniofrighetto) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16154]: getNode creates a copy, but `N0` is still the original node. It is indeed CSE'd, but the flag has been dropped by the time it gets CSE'd.

Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16154]: > but the flag has been dropped by the time it gets CSE'd.

I printed the node returned by `FindNodeOrInsertPos` and got `select_cc samesign`. The flags are not dropped. It should be fixed by https://github.com/llvm/llvm-project/pull/114650.

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16154]: Quite strange. I just tried again and confirm the node returned here:

https://github.com/llvm/llvm-project/blob/89a8c71db61282c4e7522ffda5dbee890900dca5/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp#L10321-L10322

is `i32 = select_cc t12, t28, t16, t2, setult:ch`, immediately after visiting freeze in DAGCombiner. I'm happy with your patch too.

Comment (dtcxzyw) [reply llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp:16154]: baseline: f1e1055c84bb7177f3e7d690442f3760326a0ace
```
diff --git a/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp b/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
index 7eef09e55101..b5bc6d925b41 100644
--- a/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp
@@ -16150,8 +16150,12 @@ SDValue DAGCombiner::visitFREEZE(SDNode *N) {
     R = DAG.getVectorShuffle(N0.getValueType(), SDLoc(N0), Ops[0], Ops[1],
                              SVN->getMask());
   } else {
+    errs() << "N0: ";
+    N0->dump();
     // NOTE: this strips poison generating flags.
     R = DAG.getNode(N0.getOpcode(), SDLoc(N0), N0->getVTList(), Ops);
+    errs() << "R: ";
+    R->dump();
   }
   assert(DAG.isGuaranteedNotToBeUndefOrPoison(R, /*PoisonOnly*/ false) &&
          "Can't create node that may be undef/poison!");
diff --git a/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp b/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
index d5cdd7163d79..a533ac910588 100644
--- a/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
+++ b/llvm/lib/CodeGen/SelectionDAG/SelectionDAG.cpp
@@ -10318,8 +10318,11 @@ SDValue SelectionDAG::getNode(unsigned Opcode, const SDLoc &DL, EVT VT,
     AddNodeIDNode(ID, Opcode, VTs, Ops);
     void *IP = nullptr;
 
-    if (SDNode *E = FindNodeOrInsertPos(ID, DL, IP))
+    if (SDNode *E = FindNodeOrInsertPos(ID, DL, IP)) {
+      errs() << "E: ";
+      E->dump();
       return SDValue(E, 0);
+    }
 
     N = newSDNode<SDNode>(Opcode, DL.getIROrder(), DL.getDebugLoc(), VTs);
     createOperands(N, Ops);
```
```
N0: t27: i32 = select_cc samesign t12, t28, t16, t2, setult:ch
E: t27: i32 = select_cc samesign t12, t28, t16, t2, setult:ch
R: t27: i32 = select_cc samesign t12, t28, t16, t2, setult:ch
```

**Comment by antoniofrighetto** [comment]
Closing this as favoured approach in #114650.

**Pull request closed without merging**